File: kistie-store/backend/remove-products-missing-images.js
// Script to remove products from db.sqlite that reference missing images in /images
const fs = require('fs');
const path = require('path');
const sqlite3 = require('sqlite3').verbose();

const DB_PATH = path.join(__dirname, 'db.sqlite');
const IMAGES_DIR = path.join(__dirname, '../images');

function normalize(filename) {
  return filename ? filename.trim().toLowerCase() : '';
}

function getImageFiles() {
  return fs.readdirSync(IMAGES_DIR)
    .filter(f => fs.statSync(path.join(IMAGES_DIR, f)).isFile())
    .map(f => normalize(f));
}

function main() {
  const imageFiles = new Set(getImageFiles());
  const db = new sqlite3.Database(DB_PATH);
  db.serialize(() => {
    db.all('SELECT id, name, image FROM products', (err, rows) => {
      if (err) {
        console.error('DB error:', err);
        db.close();
        return;
      }
      const toRemove = rows.filter(row => !imageFiles.has(normalize(row.image)));
      if (toRemove.length === 0) {
        console.log('No products to remove. All images exist.');
        db.close();
        return;
      }
      let removed = 0;
      const removePromises = toRemove.map(row => new Promise((resolve, reject) => {
        db.run('DELETE FROM products WHERE id = ?', [row.id], function(err2) {
          if (err2) reject(err2);
          else {
            removed++;
            resolve();
          }
        });
      }));
      Promise.all(removePromises).then(() => {
        console.log(`Removed ${removed} products with missing images.`);
        if (removed > 0) {
          console.log('Removed product IDs:');
          toRemove.forEach(row => {
            console.log(`  ID ${row.id} | Name: ${row.name} | Image: ${row.image}`);
          });
        }
        db.close();
      }).catch(e => {
        console.error('Error removing products:', e);
        db.close();
      });
    });
  });
}

if (require.main === module) {
  main();
}

File: kistie-store/flask_backend/remove_products_missing_images.py
import sqlite3
import os

# Paths
DB_PATH = os.path.join(os.path.dirname(__file__), 'db.sqlite')
IMAGES_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'React', 'public', 'images'))

conn = sqlite3.connect(DB_PATH)
c = conn.cursor()
c.execute('SELECT id, name, image FROM products')
products = c.fetchall()
removed = 0

for prod in products:
    prod_id, name, image = prod
    if not image:
        continue
    # Remove leading slashes if present
    img_file = image.replace('images/', '').lstrip('/')
    img_path = os.path.join(IMAGES_DIR, img_file)
    if not os.path.isfile(img_path):
        print(f"Deleting product '{name}' (ID: {prod_id}) - missing image: {img_file}")
        c.execute('DELETE FROM products WHERE id = ?', (prod_id,))
        removed += 1

conn.commit()
conn.close()
print(f"Removed {removed} products with missing images.")
